doc/source/data/training_ingest.rst

```diff
@@ -0,0 +1,525 @@
+    )
+
+    def train_loop_per_worker():
+        # Get an iterator to the dataset we passed in below.
+        it = session.get_dataset_shard("train")
+        for _ in range(2):
+            for batch in it.iter_batches(batch_size=128):
+                print("Do some training on batch", batch)
+
+
+    class MyCustomDataConfig(DataConfig):
+        def configure(
+            self,
+            datasets: Dict[str, Dataset],
+            world_size: int,
+            worker_handles: Optional[List[ActorHandle]],
+            worker_node_ids: Optional[List[NodeIdStr]],
+            **kwargs,
+        ) -> List[Dict[str, DataIterator]]:
+            assert len(datasets) == 1, "This example only handles the simple case"
+
+            # Configure Ray Data for ingest.
+            ctx = ray.data.DataContext.get_current()
+            ctx.execution_options = DataConfig.default_ingest_options()
+
+            # Split the stream into shards.
+            iterator_shards = datasets["train"].streaming_split(
+                world_size, equal=True, locality_hints=worker_node_ids
+            )
+
+            # Return the assigned iterators for each worker.
+            return [{"train": it} for it in iterator_shards]
+
+
+    my_trainer = TorchTrainer(
+        train_loop_per_worker,
+        scaling_config=ScalingConfig(num_workers=2),
+        datasets={"train": ds},
+        dataset_config=MyCustomDataConfig(),
+    )
+    my_trainer.fit()
+
+.. testoutput::
+    :hide:
+
+    ... 
+
+The subclass must be serializable, since it will be copied from the driver script to the driving actor of the Trainer. Its :meth:`configure <ray.train.DataConfig.configure>` method is called on the main actor of the Trainer group to create the data iterators for each worker.
+
+In general, you can use :class:`DataConfig <ray.train.DataConfig>` for any shared setup that has to occur ahead of time before the workers start iterating over data. The setup will be run at the start of each Trainer run.
+
+Performance tips
+~~~~~~~~~~~~~~~~
+
+.. _dataset_cache_performance:
+
+Caching the preprocessed Dataset
+################################
+If you are training on GPUs and have an expensive CPU preprocessing operation, this may bottleneck training throughput.
+
+If your preprocessed Dataset is small enough to fit in object store memory, the easiest thing to do is to *materialize* the preprocessed dataset in Ray object store memory, by calling :meth:`materialize() <ray.data.Dataset.materialize>` on the preprocessed dataset. This tells Ray Data to compute the entire preprocessed and pin it in the Ray object store memory. As a result, when iterating over the dataset repeatedly, the preprocessing operations do not need to be re-run. However, the trade-off is that if the preprocessed data is too large to fit into Ray object store memory, this will greatly decrease performance as data needs to be spilled to disk.
```

Comment:
```suggestion
If your preprocessed Dataset is small enough to fit in RAM, the easiest thing to do is to *materialize* the preprocessed dataset in Ray's built-in object store memory, by calling :meth:`materialize() <ray.data.Dataset.materialize>` on the preprocessed dataset. This tells Ray Data to compute the entire preprocessed and pin it in Ray object store memory. As a result, when iterating over the dataset repeatedly, the preprocessing operations do not need to be re-run.

However, if the preprocessed data is too large to fit into Ray object store memory, this will greatly decrease performance as data needs to be spilled to and read back from disk.
```

Reply:
Add a note that randomized transforms should be applied after the materialize() call?

Reply:
It needs to be able to fit in object story memory right? Which is a portion of total RAM in the cluster.

Reply:
I added this wording because "Ray object store" is jargon-y for non-core users. We could change the wording to "comfortably fit" and specify the default fraction that core uses for the object store (I think 30%).

Reply:
Clarified the default is 30%. But we do talk about object store memory throughout the configuration section. Object store memory is a foundational concept users would have to know to maximize performance.